### ISSUE TITLE
java.lang.IllegalArgumentException: signedPreKeyId MUST be greater than 0.

### ISSUE THREAD

**Issue by licaon-kter** [cmeng-git/atalk-android#47]
Can't enable OMEMO on 2.0.3 F-Droid

Mate 9, stock Pie

Self-hosted ejabberd.

**Comment by cmeng-git** [comment]
Please send me the aTalk debug log file for investigation. Meamntime ensure you enable mod_pubsub option on ejabberd.

**Comment by licaon-kter** [comment]
Conversations, Dino, Gajim and Converse.js all work fine, with OMEMO.

Debug log from where exactly? Logcat from device?

**Comment by cmeng-git** [comment]
See online help atalk.sytes.net

**Comment by licaon-kter** [comment]
Server config: https://gist.github.com/licaon-kter/edbc6629722fa1651d47924e15472bf0

Logcat when posting a message: https://gist.github.com/licaon-kter/3867b636c6d46ead58fed8e633e67cb0

**Comment by cmeng-git** [comment]
Are you ok to send me the whole debug log?

You log fragment indicates you are not connected and exception:
aTalk : java.lang.IllegalArgumentException: signedPreKeyId MUST be greater than 0

I do not have enough log info to derive any conclusion. I need to understand what went wrong prior to this happening. 



On 22 August 2019 22:13:25 GMT+08:00, Licaon_Kter <notifications@github.com> wrote:
>Server config:
>https://gist.github.com/licaon-kter/edbc6629722fa1651d47924e15472bf0
>
>Logcat when posting a message:
>https://gist.github.com/licaon-kter/3867b636c6d46ead58fed8e633e67cb0
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/cmeng-git/atalk-android/issues/47#issuecomment-523924459

--
Send from Android using XryptoMail

**Comment by licaon-kter** [comment]
I'm extracting the log now.

FYI the app freezes at random when doing actions, eg:
1. open chat with contact
2. write message
3. see big OMEMO error message
4. press title bar back button to go to roster
5. nothing happens, logcat says: https://gist.github.com/licaon-kter/23a4b0301f36b0fd3d8e3f66a413fac8
6. press back again (even wait 5 mins before pressing back)
7. see that my account is offline
8. account is online again

**Comment by licaon-kter** [comment]
Here's `atalk0.log.0`: https://gist.github.com/licaon-kter/ccac3d8369d364fb9a4d49d862412234

**Comment by cmeng-git** [comment]
Did you perform the test with Storage permission denied?

Please enabled all aTalk requested permissions before reporting any problems. You can disable them from android system settings later on after you fully understand the effects of each permission denied. Many aTalk features are inter-linked, and will have unexpected behaviors if any of these permissions is denied.

From the debug, I see the connection get disconnected on repeatedly fetching of avatar for the same user. The avatar information are stored using android file systems. I also see that the prekey index has illegal index, a corruption of database?

My advice is to uninstall aTalk to cleanup all aTalk database, reinstall aTalk, grant all permissions before doing testing and bug reporting. I would have difficulties in deriving any fruitful conclusion on debug log that have permission denied.

**Comment by licaon-kter** [comment]
Why does OMEMO need Location permission?

Why does OMEMO need Camera permission?

Why does OMEMO need Phone permission?

Why does OMEMO need Audio permission?

Why does OMEMO need Contacts permission?

Why does OMEMO need Storage permission?

**Comment by licaon-kter** [comment]
Anyway, all permissions granted, same error.

What now?

**Comment by licaon-kter** [comment]
I'd like to give you an account to test with, but I fails to find your support MUC, what's the address again?

**Comment by cmeng-git** [comment]
In my last comment, the whole problem is due to "illegal index < 0". As the debug log does not contain the full stanza exchanges. I am unable to conclude which of the system is causing problem.

Also the log indicates that there is some problems in the SSL sockect setup, and the connection kept aborted; the smack packet read continuously getting errors. This is the whole problem. I do not know exactly what is the cause of the problem. My advice is to reinstall aTalk to continue testing.

```
	at org.jivesoftware.smack.provider.Provider.parse(Provider.java:79)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:547)
	at org.jivesoftware.smack.util.PacketParserUtils.parseStanza(PacketParserUtils.java:116)
	at org.jivesoftware.smack.AbstractXMPPConnection.parseAndProcessStanza(AbstractXMPPConnection.java:1252)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$500(XMPPTCPConnection.java:136)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:936)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.access$300(XMPPTCPConnection.java:882)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:902)
	at java.lang.Thread.run(Thread.java:784)
```

aTalk supports sending of audio clips, pictures and locations etc during all chat sessions.
If you have gateway setup for telephony at your server, phone/contact info is required in this case.

If you would like me to perform the test for OMEMO chat using your server, please provide two test accounts.

**Comment by licaon-kter** [comment]
Where is your support MUC?

I've reinstalled it every time I tested, same error.

**Comment by cmeng-git** [comment]
You can  send the test accounts info directly to my mail cmeng.gm@gmail.comn

**Comment by licaon-kter** [comment]
@cmeng-git Please join the Conversations MUC so we can talk or close this issue.

Dogfooding...   ¯\_(ツ)_/¯

**Comment by licaon-kter** [comment]
BTW, does aTalk have any key number limits? 

This account is used for Converse.js testing too, so it accumulated a lot of keys (>25) that [I can't actually delete](https://github.com/conversejs/converse.js/issues/1666) _easily_.

**Comment by cmeng-git** [comment]
Now I understand what you meant by MUC support. I am the lone individual whom supports the development of aTalk. There is no purpose for me to setup an MUC talking to oneself. 
Furthermore I don't believe we are able to conclude or derive any answer to problems encountered via conference. I need the actual detailed debug log to find the root cause.

It is not appropriate and impolite to discuss aTalk issues in conversations conference. If you really need my assistance to resolve the problem, the quicker way is to provide me with two test accounts so I can logged into your server. You can remove the accounts in a week or two after the problem has resolved.

**Comment by cmeng-git** [comment]
https://github.com/cmeng-git/atalk-android/issues/47#issuecomment-524215223

aTalk v2.0.3 has actually implemented "Purge unused identities" 

[[20190716] When do I have to perform "Regenerate Omemo identity"?](https://atalk.sytes.net/atalk/faq.html#omemo_04).

But still having problems,  Please refer to the links for more info:

a. [Smack 4.4.0: Regenerate OMEMO Identity Implementation ](https://discourse.igniterealtime.org/t/smack-4-4-0-regenerate-omemo-identity-implementation/85558/4)
b. [PubSub: Delete an Item from a Node published returns error - Node not found even though it exists #2974](https://github.com/processone/ejabberd/issues/2974)

**Comment by licaon-kter** [comment]
> It is not appropriate and impolite to discuss aTalk issues in conversations conference.

Not discuss in the MUC, but contact you personally to give you the credentials.

Am I lost in translation here?

**Comment by licaon-kter** [comment]
> aTalk v2.0.3 has actually implemented "Purge unused identities"

Great, but not sure if related. aTalk should create it's own device keys and not touch the others, by other clients.

**Comment by cmeng-git** [comment]
aTalk OMEMO device only have information on its own deviceKeys, and are only capable of deleting its own inactive omemoDevices.

> aTalk should cope with whatever identities/keys/prekeys the user has.
Presently there is no problem for aTalk to handle its own keys. But note that having inactive omemoDevices keys do load the servers and the android devices itself. Like what you have mentioned, the 25 Keys are actually put a load on server; imaging with thousands users having multiple keys with all inactive except one.

I only use aTalk xmpp client for testing, but no for normal use. You can try to send the info to
swordfish@atalk.sytes.net. I will log onto this account to receive the info.

**Comment by licaon-kter** [comment]
> I only use aTalk xmpp client for testing, but no for normal use
That's an odd thing to say/do


Aaaanyway, I've added you...waiting for you to ack and get the credentials.

**Comment by cmeng-git** [comment]
I have just installed the two accounts on my Note-3 and Note-8 android devices.
I am able to send and receive OMEMO messages without problem. I have forwarded the screenshot to you. So there is no problem for aTalk to work on your server, as least from my test results so far.

1. May I know in your earlier test setup that is having problem, was it a 1-1 chat or conference chat.
2. Also are all your test accounts reside on the same test server that I am using now.

Are you able to send omemo messages to the atalkuser1 from the accounts that you are having problem with. So that I can see what is actually being received on my side.

**Comment by licaon-kter** [comment]
1:1, same server

After I added those 25 identities to User1 I've repro the issue yet again between User1 and User2: User1 with aTalk, User2 with Converse.

Message from User2 reach User1 encrypted, user1 errors out when transmitting.

**Comment by cmeng-git** [comment]
After more testing, below is the root cause of the problem:

The prekeys bundle which was generated by another xmpp client (non-aTalk), actually contains an illegal prekey i.e. \<preKeyPublic preKeyId='0'> i.e. with index=0. This is being rejected by smack OmemoBundleElement handler and throws an exception. Hence aTalk is unable to proceed with the omemo message sending; and delay this exception error message to the chat window. This is the error message that you have observed.

`IllegalArgumentException: signedPreKeyId MUST be greater than 0.
`

In conversations implementation, it just ignores this erroneous omemo device and strip this device from its recipient list  with an message displays on chat to that effect:
`"Message was not encrypted for this device." `

From testing it also indicates that conversation client is not the client that is causing the problem, but another xmpp client that was used to generate the prekey bundle during testing.

I will relay this info to the smack OMEMO team for advice.

```
============== aTalk debug log ===================
08-24 08:37:55.487 32673-4267/org.atalk.android D/SMACK: RECV (0): 
    <iq xml:lang='en' to='user2@myserver.ltd/atalk' from='user2@myserver.ltd' type='result' id='2YL2F-321'>
      <pubsub xmlns='http://jabber.org/protocol/pubsub'>
        <items node='eu.siacs.conversations.axolotl.bundles:8870'>
          <item id='61E8C7EEE7D60'>
            <bundle xmlns='eu.siacs.conversations.axolotl'>
              <signedPreKeyPublic signedPreKeyId='0'>
                Bez7xfU8lVaXxjfiWuFVfoCdtHjGBzI4nays/p8dzAxe
              </signedPreKeyPublic>
              <signedPreKeySignature>
                +94sRJNK+ysxVVuqYVXQT0652Dm26xTt4C7EvFJ2oc7eY6kFVtt8G3fyhKGefgqFc2Tc5eOOyq/4+1WJim6Zjg==
              </signedPreKeySignature>
              <identityKey>
                BcuPpYfkfn3UnQ1mf0JUKUFLBknz6xv5QFWVoRxHc1MH
              </identityKey>
              <prekeys>
                <preKeyPublic preKeyId='0'>
                  BcgjrJk4KR9ghqGLoNEeccBm0NTcvBKebKtuAH4Tlk5c
                </preKeyPublic>
                <preKeyPublic preKeyId='1'>
                  BXmjl6Tuv+9+Osy7dQtCYq/fb+K3ssbJywimJgmIfER5
.....
```

**Comment by cmeng-git** [comment]
Currently, the above problem can be restored to working condition using aTalk.
a. Login in aTalk with the user account e.g. user2@myserver.ltd, whom prekeys bundles contains index='0'. From log, the device is \<items node='eu.siacs.conversations.axolotl.bundles:8870'/>
b. Perform the "Regenerate OMEMO identities" 
c. Then exit and restart aTalk.

For the next aTalk v.2.0.4 release, step (c) will not be required.  "Regenerate OMEMO identities"  will not required to restart aTalk.

**Comment by cmeng-git** [comment]
fixed in aTalk version 2.0.4 - changed to allow index=0
